File: LeetCode/Python/1137_N-th_Tribonacci_Number.py
# 1137. N-th Tribonacci Number
# https://leetcode.com/problems/n-th-tribonacci-number/


class Solution:
    def tribonacci(self, n: int) -> int:
        if n == 0:
            return 0
        elif n == 1:
            return 1
        elif n == 2:
            return 1
        a, b, c, d = 0, 1, 1, 0
        while n - 2 > 0:
            d = a + b + c
            a = b
            b = c
            c = d
            n -= 1
        return d

File: LeetCode/JavaScript/1137-N-th-Tribonacci-Number.js
// 1137. N-th Tribonacci Number
// https://leetcode.com/problems/n-th-tribonacci-number/

/**
 * @param {number} n
 * @return {number}
 */

var tribonacci = function (n) {
  const memo = {
    0: 0,
    1: 1,
    2: 1,
  };

  function helper(n) {
    if (memo[n] !== undefined) {
      return memo[n];
    }

    memo[n] = helper(n - 1) + helper(n - 2) + helper(n - 3);

    return memo[n];
  }

  return helper(n);
};
